ipld/amt/src/iter.rs

```diff
@@ -0,0 +1,559 @@
+}
+
+pub struct Iter<'a, V, BS, Ver> {
+    stack: Vec<IterStack<'a, V>>,
+    blockstore: BS,
+    bit_width: u32,
+    ver: PhantomData<Ver>,
+    key: usize,
+}
+
+pub struct IterStack<'a, V> {
+    pub(crate) node: Option<&'a Node<V>>,
+    pub(crate) idx: usize,
+}
+
+impl<'a, V, BS, Ver> Iterator for Iter<'a, V, BS, Ver>
+where
+    BS: Blockstore,
+    V: Serialize + DeserializeOwned,
+{
+    type Item = Result<(usize, &'a V), crate::Error>;
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            let stack = self.stack.last_mut()?;
+            match stack.node {
+                Some(Node::Leaf { vals }) => {
+                    while stack.idx < vals.len() {
+                        match vals[stack.idx] {
+                            Some(ref v) => {
+                                stack.idx += 1;
+                                self.key += 1;
+                                return Some(Ok((self.key - 1, v)));
+                            }
+                            None => {
+                                stack.idx += 1;
+                                self.key += 1;
+                            }
+                        }
+                    }
+                    self.stack.pop();
+                }
+                Some(Node::Link { links }) => {
+                    if stack.idx < links.len() {
+                        let link = &links[stack.idx];
+                        match link {
+                            Some(Link::Cid { cid, cache }) => {
+                                match cache.get_or_try_init(|| {
+                                    self.blockstore
+                                        .get_cbor::<CollapsedNode<V>>(cid)?
+                                        .ok_or_else(|| Error::CidNotFound(cid.to_string()))?
+                                        .expand(self.bit_width)
+                                        .map(Box::new)
+                                }) {
+                                    Ok(node) => {
+                                        stack.idx += 1;
+                                        self.stack.push(IterStack {
+                                            node: Some(node.as_ref()),
+                                            idx: 0,
+                                        });
+                                    }
+                                    Err(e) => return Some(Err(e)),
+                                }
+                            }
+                            Some(Link::Dirty(node)) => {
+                                stack.idx += 1;
+                                self.stack.push(IterStack {
+                                    node: Some(node.as_ref()),
+                                    idx: 0,
+                                });
+                            }
+                            None => {
+                                stack.idx += 1;
+                                self.key += 2_usize.pow(self.bit_width);
```

Comment:
This depends on the height. You need to call `nodes_for_height` with the `bit_width` and `root.height - self.stack.len()` (plus or minus 1-2, you'll need to test this carefully).

Reply:
Ah, that's not quite sparse enough. Try changing 64 to 500.

Basically, the current logic works for 2 levels, but not more.

Reply:
Good catch. Fixed.